### ISSUE TITLE
needless borrow does not trigger on block trailing expression

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust-clippy#1726]
```rust
    let x = 5;
    let foo = match 42 {
        44 => &x,
        45 => {
            println!("foo");
            &&x // does not trigger needless_borrow
        },
        46 => &&x, // triggers needless_borrow
        _ => panic!(),
    };
```

**Comment by oli-obk** [comment]
Ok... so this doesn't trigger on any expression where the location of the autoderef is disconnected from the location of the address-taking. In the match example, the issue is probably in the expansion of match arms.

e.g.

```rust
    let intermediate = &&5;
    let b: &i32 = intermediate; // doesn't lint
    let c: &i32 = &&5; // lints
```

This just became much harder to solve.

Any complaints about a lint on locals with `&&T` type which are only ever used by (auto-)derefing? The suggestion would be to deref at the initialization site instead of at each use site.
